tesseract/messaging/fees/src/lib.rs

```diff
@@ -451,45 +453,127 @@ impl TransactionPayment {
 			return Ok(Default::default());
 		}
 
-		let requests = request_entries.iter().filter_map(|data| {
-			// Get request commitment keys on source chain
-			let hash = H256::from_slice(&hex::decode(data.hash.clone()).ok()?);
-			let source_key = source.request_commitment_full_key(hash);
-			//Get request receipt keys on dest chain
-			let dest_key = dest.request_receipt_full_key(hash);
-			Some((Key::Request(hash), source_key, dest_key))
-		});
+		let req_results: Vec<_> = stream::iter(request_entries)
+			.then(|data| {
+				let source = source.clone();
+				let dest = dest.clone();
+				async move {
+					// Get request commitment keys on source chain
+					let hash = match hex::decode(data.hash.clone()) {
+						Ok(bytes) => H256::from_slice(&bytes),
+						Err(_) => return None,
+					};
+
+					match source.query_request_fee_metadata(hash).await {
+						Ok(fee) if fee.is_zero() => Some(Err(Key::Request(hash))),
+						Ok(_) => {
+							let source_key = source.request_commitment_full_key(hash);
+							//Get request receipt keys on dest chain
+							let dest_key = dest.request_receipt_full_key(hash);
+							Some(Ok((Key::Request(hash), source_key, dest_key)))
+						}
+						Err(err) => {
```

Comment:
This implementation is a bit unreadable. I think it's better if you just query the fee metadata inside this block and push keys with zero fees into the `keys_to_delete` array
https://github.com/polytope-labs/hyperbridge/blob/0f37db9c89cd331a3608566d7f7ef02352338320/tesseract/messaging/fees/src/lib.rs#L477